src/V3Const.cpp

```diff
@@ -590,6 +590,86 @@ class ConstVisitor : public AstNVisitor {
         return false;
     }
 
+    // 'out' will have those bits set, which are known to be set
+    void setBits(V3Number& out, AstNode* nodep) {
+        // @wsnyder Do I need some sizedness test here, or are they already well formed?
+        if (const AstConst* const np = VN_CAST_CONST(nodep, Const)) {
+            out.opAssign(np->num());
+        } else if (const AstOr* const np = VN_CAST_CONST(nodep, Or)) {
+            // @wsnyder Do I need to check operands are sized and have the same width?
+            V3Number lb(nodep, np->lhsp()->widthMin());
+            V3Number rb(nodep, np->rhsp()->widthMin());
+            setBits(lb, np->lhsp());
+            setBits(rb, np->rhsp());
+            out.opOr(lb, rb);
+        } else {
+            // Pretend we don't know anything
+            out.setAllBits0();
+        }
+    }
+
+    // 'out' will have those bits set, which are known to be clear
+    void clrBits(V3Number& out, AstNode* nodep) {
+        // @wsnyder Do I need some sizedness test here, or are they already well formed?
+        if (const AstConst* const np = VN_CAST_CONST(nodep, Const)) {
+            out.opNot(np->num());
+        } else if (const AstOr* const np = VN_CAST_CONST(nodep, Or)) {
+            // @wsnyder Do I need to check operands are sized and have the same width?
+            V3Number lb(nodep, np->lhsp()->widthMin());
+            V3Number rb(nodep, np->rhsp()->widthMin());
+            clrBits(lb, np->lhsp());
+            clrBits(rb, np->rhsp());
+            out.opAnd(lb, rb);
+        } else if (const AstVarRef* const np = VN_CAST_CONST(nodep, VarRef)) {
+            out.setAllBits0();
+            for (int bit = np->varp()->widthMin(); bit < nodep->widthMin(); ++bit) {
+                out.setBit(bit, 1);
+            }
+        } else {
+            // Pretend we don't know anything
+            out.setAllBits0();
+        }
+    }
+
+    bool checkKnownBits(AstNodeBiop* nodep) {
+        // Check if the sides have enough known bits to to determine the result
+        // This mostly helps to squash compiler warnings so we are not doing it
+        // too excessively.
+        const int width = nodep->lhsp()->widthMin();
+        // @wsnyder Is this check necessary, or are they already well formed?
+        if (width == 0 || width != nodep->lhsp()->widthMin()) return false;
```

Comment:
Generally they should be well formed, but generally in V3Const I play it safe and always check. You shouldn't need to use widthMin, width() should work for all boolean ops (shifts & casts are somewhat special).  Ditto elsewhere.

```suggestion
        if (width == 0 || width != nodep->lhsp()->width()) return false;
```

Reply:
I'm mostly using `widthMin` as I used the local `width` below to pass to the V3Number instances. 

Am I right in thinking that `width() != widthMin()` if and only if the variable has already been allocated to a C type, in which case `width()` is the width of the C type, and `widthMin()` is the width of the Verilog type stored in it? If so then I think I need to pass `widthMin()` to the V3Numbers below.